### PR TITLE
[FLINK-27965][table-planner] Fix EXPLAIN PLAN produces wrong query sc…

### PR DESCRIPTION
**Pull request by LadyForest**
## What is the purpose of the change

This pull request fixes the issue that the `EXPLAIN PLAN FOR INSERT` with static partition spec produces the wrong query schema.

During the conversion phase, the query will be validated and rewritten by `PreValidateReWriter`, and `PreValidateReWriter#appendPartitionAndNullsProjects` will perform the following rewrite(https://github.com/apache/flink/blob/master/flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/calcite/PreValidateReWriter.scala#L89).
```sql
-- before rewrite
insert into A partition(a='11',
 c='22') select b from B

-- after rewrite
insert into A partition(a='11', c='22') select cast('11' as tpe1), b, cast('22' as tpe2) from B
```
However, the `SqlRichExplain` are first validated as a whole and then got its statement validated again, thus the insert query is rewritten twice, which causes the issue.

## Brief change log
  - For `SqlRichExplain`, let`FlinkPlannerImpl` validate the insert's source instead of insert.

## Verifying this change

This change added tests and can be verified as follows:

  - Add `TableEnvironmentTest#testExecuteSqlWithExplainInsertStaticPartition`
  - Add it case in `table.q`
 

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/calcite/FlinkPlannerImpl.scala [hunk 1]
```diff
@@ -149,7 +149,20 @@ class FlinkPlannerImpl(
       }
       sqlNode match {
         case richExplain: SqlRichExplain =>
-          richExplain.setOperand(0, validate(richExplain.getStatement))
+          val validatedStatement = richExplain.getStatement match {
+            // only validate source here
+            case insert: RichSqlInsert =>
+              if (insert.isUpsert) {
```
Comment (JingsongLi) [inline flink-table/flink-table-planner/src/main/scala/org/apache/flink/table/planner/calcite/FlinkPlannerImpl.scala:155]: Extract a method for reusing?

**Review by JingsongLi — COMMENTED** [review]
Thanks for the contribution, left one comment.

**Comment by LadyForest** [comment]
@flinkbot run azure

**Review by JingsongLi — APPROVED** [review]
Looks good to me! Thanks for the update.

**Pull request merged**